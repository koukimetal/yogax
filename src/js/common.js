import React from "react";
import {Record, Set} from 'immutable';

const PartState = {
    NONE: 0,
    OCCUPIED: 1,
    PENDING: 2,
};

Object.freeze(PartState);

class Part extends Record({state: PartState.NONE, player: -1, groupId: -1}, 'Part') {}
const Coordinate = Record({x: 0, y: 0}, 'Coordinate');
class PendingProperty extends Record({x: -1, y: -1, shape: null}, 'PendingProperty') {}

const BOARD_SIZE = 14;
const PART_LENGTH = 5;
const PART_WIDTH = 40;
const VEC_CHOOSE = Object.freeze([-1, 0, 1]);
const START_POINT = Set().add(new Coordinate({x: 4, y: 4})).add(new Coordinate({x: 9, y: 9}));
const DX = [0, 1, 0, -1];
const DY = [-1, 0, 1, 0];

const locatableForChoice = (baseX, baseY, seq, field) => {
    for (let i = 0; i < seq.length; i++) {
        const co = seq[i];
        for (let dx of VEC_CHOOSE) {
            for (let dy of VEC_CHOOSE) {
                const cx = co.x + dx + baseX;
                const cy = co.y + dy + baseY;

                if (cx < 0 || field.length <= cx || cy < 0 || field.length <= cy) {
                    return false;
                }

                if (field[cy][cx].state === PartState.OCCUPIED) {
                    return false;
                }
            }
        }
    }
    return true;
};

const inBorder = (x, y, size) => {
    return 0 <= x && x < size && 0 <= y && y < size;
};

const getStrokeDasharray = (x, y, field, edgeLength) => {
    if (field[y][x].player === -1) {
        return [];
    }

    const connection = [];
    for (let i = 0; i < DX.length; i++) {
        const cx = x + DX[i];
        const cy = y + DY[i];

        if (!inBorder(cx, cy, field.length)) {
            continue;
        }

        if (field[cy][cx].groupId === field[y][x].groupId) {
            connection.push(i);
        }
    }

    let draw = true;
    const dashLine = [0].concat(Array(5).fill(edgeLength / 5)).concat([0]);
    const result = [];
    for (let i = 0; i < DX.length;) {
        if (connection.includes(i) && draw) {
            result.push(...dashLine);
            i++;
        } else if (!connection.includes(i) && draw) {
            result.push(edgeLength);
            i++;
        } else {
            result.push(0);
        }
        draw = !draw;
    }

    // compress
    const optimized = [];
    for (let i = 0; i < result.length;) {
        if ( // 20, 0, 20 -> 40
        (optimized.length > 0 && i + 1 < result.length) &&
        optimized[optimized.length - 1] > 0 &&
        optimized[optimized.length - 1] % edgeLength === 0 &&
        result[i] === 0 &&
        result[i + 1] === edgeLength
        ) {
            optimized[optimized.length - 1] += result[i + 1];
            i += 2;
        } else if ( // 0, 0, 0 -> 0
        (optimized.length > 0 && i + 1 < result.length) &&
        optimized[optimized.length - 1] === 0 &&
        result[i] === 0 &&
        result[i + 1] === 0
        ) {
            i += 2;
        } else {
            optimized.push(result[i]);
            i++;
        }
    }

    return optimized;
};

const getField = (size) => {
    const field = new Array(size);
    for (let i = 0; i < size; i++) {
        field[i] = new Array(size);
        for (let j = 0; j < size; j++) {
            field[i][j] = new Part();
        }
    }
    return field;
};

export {
    Part,
    PartState,
    Coordinate,
    BOARD_SIZE,
    PART_LENGTH,
    PART_WIDTH,
    getStrokeDasharray,
    START_POINT,
    inBorder,
    locatableForChoice,
    DX,
    DY,
    getField,
}