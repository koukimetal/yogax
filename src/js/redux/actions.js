
const putPart = (player, groupId, pendingParts) => ({
    type: 'PUT_PART',
    player,
    groupId,
    pendingParts,
});

const selectPart = (groupId) => ({
    type: 'SELECT_PART',
    groupId,
});

const moveCursor = (cursor) => ({
    type: 'MOVE_CURSOR',
    cursor,
});

const operateShape = (operation) => ({
    type: 'OPERATE_SHAPE',
    operation,
});

const pass = () => ({
    type: 'PASS',
});

export {
    putPart,
    selectPart,
    moveCursor,
    operateShape,
    pass,
};