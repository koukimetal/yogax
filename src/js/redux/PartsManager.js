import {generateAllShapes} from "../shape";
import {fromJS} from 'immutable';
import {PART_LENGTH} from "../common";

class PartsManager {
    constructor(size, numOfPlayer) {
        const initialShapes = generateAllShapes(size);
        const groupIdToShape = [];
        const groupIdToPlayer = [];

        for (let player = 0; player < numOfPlayer; player++) {
            initialShapes.forEach(shape => {
                groupIdToShape.push(shape);
                groupIdToPlayer.push(player);
            });
        }
        this.groupIdToShape = fromJS(groupIdToShape);
        this.groupIdToPlayer = fromJS(groupIdToPlayer);

    }

    getPlayer(groupId) {
        return this.groupIdToPlayer.get(groupId);
    }

    getShape(groupId) {
        return this.groupIdToShape.get(groupId);
    }

    getNumOfGroupId() {
        return this.groupIdToShape.size;
    }
}

const PartsManagerInstance = new PartsManager(PART_LENGTH, 2);

export default PartsManagerInstance;