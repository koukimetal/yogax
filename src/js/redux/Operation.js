import React from "react";
import {operateShape as getOperateAction, pass as getPassAction} from "./actions";
import {connect} from "react-redux";

const Operation = ({
   player,
   alreadyPut,
   operateShape,
   pass
}) => {
    return (
        <div>
            <div>
                Player: {player + 1}
            </div>
            <button onClick={() => operateShape('rotateRight')}>Rotate Right</button>
            <button onClick={() => operateShape('rotateLeft')}>Rotate Left</button>
            <button onClick={() => operateShape('flip')}>Flip right top to left bottom</button>
            <button onClick={() => pass()}>Pass</button>

            <div>
                Score 1: {alreadyPut.get(0)}, Score 2: {alreadyPut.get(1)}
            </div>
        </div>
    );
};

const mapStateToProps = state => ({
    player: state.player,
    alreadyPut: state.alreadyPut,
});

const mapDispatchToProps = dispatch => ({
    operateShape: (operation) => {
        dispatch(getOperateAction(operation));
    },
    pass: () => {
        dispatch(getPassAction());
    }
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Operation);
