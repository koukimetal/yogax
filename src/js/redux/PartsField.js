import {PART_WIDTH, BOARD_SIZE, getStrokeDasharray} from "../common";
import {selectPart as getSelectPartAction} from "./actions";
import { connect } from 'react-redux';
import React from "react";
import partsManager from "./PartsManager";

const PartsFieldComponent = ({
    player,
    field,
    selectedGroupId,
    selectPart
}) => {
    const fieldJS = field.toJS();
    let canvas = [];
    for (let y = 0; y < fieldJS.length; y++) {
        for (let x = 0; x < fieldJS[y].length; x++) {
            const strokeDasharray = getStrokeDasharray(x, y, fieldJS, PART_WIDTH/2);
            let fillColor = 'gray';

            if (fieldJS[y][x].groupId === selectedGroupId) {
                fillColor = 'green';
            } else if (fieldJS[y][x].player === 0) {
                fillColor = 'blue';
            } else if (fieldJS[y][x].player === 1) {
                fillColor = 'red';
            }

            canvas.push(<rect
                key={y * fieldJS.length + x}
                x={x * PART_WIDTH/2}
                y={y * PART_WIDTH/2}
                width={PART_WIDTH/2}
                height={PART_WIDTH/2}
                onClick={() => selectPart(player, fieldJS[y][x].groupId)}
                style={{
                    fill: fillColor,
                    stroke: 'white',
                    strokeWidth: 2,
                    strokeDasharray,
                }}
            />);
        }
    }

    return (
        <svg width={PART_WIDTH * BOARD_SIZE} height={PART_WIDTH * BOARD_SIZE}>
            {canvas}
        </svg>
    );
};

const mapStateToProps = state => ({
    player: state.player,
    field: state.partsField,
    selectedGroupId: state.selectedGroupId,
});

const mapDispatchToProps = dispatch => ({
    selectPart: (player, groupId) => {
        if (groupId >= 0) {
            const partsPlayer = partsManager.getPlayer(groupId);
            if (player === partsPlayer) {
                dispatch(getSelectPartAction(groupId));
            }
        }
    }
});

export {
    PartsFieldComponent,
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(PartsFieldComponent);
