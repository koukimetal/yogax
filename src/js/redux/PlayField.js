import {Coordinate, PART_WIDTH, getStrokeDasharray, BOARD_SIZE, START_POINT, inBorder, PartState, DX, DY} from "../common";
import React from "react";
import {Set} from 'immutable';
import {putPart as getPutPartAction, moveCursor as getMoveCursorAction} from "./actions";
import {connect} from "react-redux";


const getPendingParts = (shape, cursor) => {
    return shape.seq.map(co => new Coordinate({x: co.x + cursor.x, y: co.y + cursor.y}));
};

const drawable = (fieldJS, seq) => {
    let valid = true;
    seq.forEach(co => {
        valid = valid && inBorder(co.x, co.y, BOARD_SIZE) && fieldJS[co.y][co.x].state === PartState.NONE;
    });
    return valid;
};

const PlayField = ({
    turn,
    player,
    selectedGroupId,
    selectedShape,
    field,
    cursor,

    moveCursor,
    putPart
}) => {
    const fieldJS = field.toJS();

    let pendingParts = Set();
    if (cursor && selectedShape) {
        const pendingPartsCandidate = getPendingParts(selectedShape, cursor);
        if (drawable(fieldJS, pendingPartsCandidate)) {
            pendingParts = pendingPartsCandidate;
        }
    }

    let canvas = [];
    for (let y = 0; y < BOARD_SIZE; y++) {
        for (let x = 0; x < BOARD_SIZE; x++) {
            const co = new Coordinate({x, y});

            let fillColor = 'gray';

            if (START_POINT.includes(co)) {
                fillColor = 'orange';
            }

            if (pendingParts.includes(co)) {
                fillColor = 'purple';
            }

            const strokeDasharray = getStrokeDasharray(x, y, fieldJS, PART_WIDTH);

            if (fieldJS[y][x].player === 0) {
                fillColor = 'blue';
            } else if (fieldJS[y][x].player === 1) {
                fillColor = 'red';
            }

            canvas.push(<rect
                key={y * BOARD_SIZE + x}
                x={x * PART_WIDTH}
                y={y * PART_WIDTH}
                width={PART_WIDTH}
                height={PART_WIDTH}
                onMouseOver={() => moveCursor(fieldJS, selectedShape, x, y)}
                onClick={() => putPart(turn, player, field, selectedGroupId, pendingParts)}
                style={{
                    fill: fillColor,
                    stroke: 'white',
                    strokeWidth: 2,
                    strokeDasharray,
                }}
            />);
        }
    }

    return (
        <svg width={PART_WIDTH * BOARD_SIZE} height={PART_WIDTH * BOARD_SIZE}>
            {canvas}
        </svg>
    );
};

const isLocatable = (turn, player, field, pendingParts) => {
    if (pendingParts.size === 0) { // not chosen
        return;
    }

    // check locatable
    let playFieldJS = field.toJS();
    let locatable = true;

    // check not sharing edges
    pendingParts.forEach(co => {
        if (playFieldJS[co.y][co.x].state !== PartState.NONE) {
            locatable = false;
        }
        for (let i = 0; i < DX.length; i++) {
            const nx = co.x + DX[i];
            const ny = co.y + DY[i];
            if (!inBorder(nx, ny, BOARD_SIZE)) {
                continue;
            }
            if (playFieldJS[ny][nx].player === player) {
                locatable = false;
            }
        }
    });

    if (turn < 2) {
        // check including a start point
        let includeStart = false;
        pendingParts.forEach(co => {
            includeStart = includeStart || START_POINT.includes(co);
        });
        locatable = locatable && includeStart;
    } else {
        // check sharing at least one corner
        const requiredSameColorPlaces = [];
        pendingParts.forEach(co => {
            const connectedVectors = [];
            for (let i = 0; i < DX.length; i++) {
                const nco = new Coordinate({x: co.x + DX[i], y: co.y + DY[i]});
                if (!inBorder(nco.x, nco.y, BOARD_SIZE) || !pendingParts.includes(nco)) {
                    continue;
                }

                connectedVectors.push({x: DX[i], y: DY[i]});
            }

            if (connectedVectors.length === 0) {
                requiredSameColorPlaces.push(new Coordinate({x: co.x - 1, y: co.y - 1}));
                requiredSameColorPlaces.push(new Coordinate({x: co.x + 1, y: co.y + 1}));
                requiredSameColorPlaces.push(new Coordinate({x: co.x - 1, y: co.y + 1}));
                requiredSameColorPlaces.push(new Coordinate({x: co.x + 1, y: co.y - 1}));
            } else if (connectedVectors.length === 1) {
                const cv = connectedVectors[0];
                if (cv.x !== 0) {
                    requiredSameColorPlaces.push(new Coordinate({x: co.x + cv.x * -1, y: co.y + 1}));
                    requiredSameColorPlaces.push(new Coordinate({x: co.x + cv.x * -1, y: co.y - 1}));
                } else {
                    requiredSameColorPlaces.push(new Coordinate({x: co.x + 1, y: co.y + cv.y * -1}));
                    requiredSameColorPlaces.push(new Coordinate({x: co.x - 1, y: co.y + cv.y * -1}));
                }
            } else if (connectedVectors.length === 2) {
                const cv1 = connectedVectors[0];
                const cv2 = connectedVectors[1];
                requiredSameColorPlaces.push(new Coordinate({x: co.x - cv1.x - cv2.x, y: co.y - cv1.y - cv2.y}));
            }
        });

        let satisfy = false;
        requiredSameColorPlaces.forEach(co => {
            if (!inBorder(co.x, co.y, BOARD_SIZE)) {
                return;
            }

            if (playFieldJS[co.y][co.x].player === player) {
                satisfy = true;
            }
        });

        locatable = locatable && satisfy;
    }

    return locatable;
};

const mapStateToProps = state => {
    return {
        turn: state.turn,
        player: state.player,
        selectedGroupId: state.selectedGroupId,
        selectedShape: state.selectedShape,
        field: state.playField,
        cursor: state.cursor,
    }

};

const mapDispatchToProps = dispatch => ({
    putPart: (turn, player, field, selectedGroupId, pendingParts) => {
        if (isLocatable(turn, player, field, pendingParts)) {
            dispatch(getPutPartAction(player, selectedGroupId, pendingParts));
        }
    },
    moveCursor: (fieldJS, selectedShape, x, y) => {
        if (!selectedShape) {
            return;
        }
        const nextPendingParts = getPendingParts(selectedShape, new Coordinate({x, y}));
        if (drawable(fieldJS, nextPendingParts)) {
            dispatch(getMoveCursorAction(new Coordinate({x, y})));
        }
    }
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(PlayField);
