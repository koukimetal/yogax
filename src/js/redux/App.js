import React from 'react';

import Operation from './Operation';
import PartsField from './PartsField';
import PlayField from './PlayField';

const App = () => (
    <div>
        <PlayField />
        <PartsField />
        <Operation />
    </div>
);
export default App;

