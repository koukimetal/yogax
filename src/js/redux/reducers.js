import {Part, PartState, BOARD_SIZE, PART_LENGTH, locatableForChoice} from "../common";
import {fromJS, List} from 'immutable';
import partsManager from "./PartsManager";
import { combineReducers } from 'redux';

const getField = (size) => {
    const field = new Array(size);
    for (let i = 0; i < size; i++) {
        field[i] = new Array(size);
        for (let j = 0; j < size; j++) {
            field[i][j] = new Part();
        }
    }
    return field;
};

const playField = (state, action) => {
    if (typeof state === 'undefined') {
        return fromJS(getField(BOARD_SIZE));
    }

    if (action.type === 'PUT_PART') {
        const pendingParts = action.pendingParts;

        const field = state.toJS();

        pendingParts.forEach(co => {
            field[co.y][co.x] = new Part({state: PartState.OCCUPIED, player: action.player, groupId: action.groupId});
        });

        return fromJS(field);
    }

    return state;
};

const partsField = (state, action) => {
    if (typeof state === 'undefined') {
        let partsField = getField(BOARD_SIZE * 2);

        const groupIdMax = partsManager.getNumOfGroupId();

        let groupId = 0;
        for (let y = 0; y < partsField.length && groupId < groupIdMax; y++) {
            for (let x = 0; x < partsField[y].length && groupId < groupIdMax; x++) {
                const seq = partsManager.getShape(groupId).seq;
                if (locatableForChoice(x, y, seq.toJS(), partsField)) {
                    seq.forEach(co => {
                        partsField[co.y + y][co.x + x] = new Part({
                            state: PartState.OCCUPIED,
                            player: partsManager.getPlayer(groupId),
                            groupId
                        });
                    });
                    groupId++;
                }
            }
        }

        return fromJS(partsField);
    }

    if (action.type === 'PUT_PART') {
        const field = state.toJS();

        for (let y = 0; y < field.length; y++) {
            for (let x = 0; x < field[y].length; x++) {

                if (field[y][x].groupId === action.groupId) {
                    field[y][x] = new Part();
                }
            }
        }

        return fromJS(field);
    }

    return state;
};

const selectedShape = (state = null, action) => {
    if (action.type === 'SELECT_PART') {
        return partsManager.getShape(action.groupId);
    } else if (action.type === 'PUT_PART' || action.type === 'PASS') {
        return null;
    } else if (action.type === 'OPERATE_SHAPE' && state) {
        if (action.operation === 'rotateRight') {
            return state.rotate90();
        } else if (action.operation === 'rotateLeft') {
            return state.rotate90().rotate90().rotate90();
        } else if (action.operation === 'flip') {
            return state.flip();
        }
    }

    return state;
};

const cursor = (state = null, action) => {
    if (action.type === 'PUT_PART' || action.type === 'SELECT_PART' || action.type === 'PASS') {
        return null;
    } else if (action.type === 'MOVE_CURSOR') {
        return action.cursor;
    }

    return state;
};

const selectedGroupId = (state = -2, action) => {
    if (action.type === 'PUT_PART' || action.type === 'PASS') {
        return -2;
    } else if (action.type === 'SELECT_PART') {
        return action.groupId;
    }

    return state;
};

const player = (state = 0, action) => {
    if (action.type === 'PUT_PART' || action.type === 'PASS') {
        return (state + 1) % 2;
    }

    return state;
};

const turn = (state = 0, action) => {
    if (action.type === 'PUT_PART') {
        return state + 1;
    }

    return state;
};

const alreadyPut = (state, action) => {
    if (typeof state === 'undefined') {
        return List().push(0).push(0);
    }

    if (action.type === 'PUT_PART') {

        const player = action.player;
        const shapeSize = partsManager.getShape(action.groupId).seq.size;

        return state.withMutations(list => {
            const original = list.get(player);
            list.set(player, original + shapeSize);
        });
    }

    return state;
};

export default combineReducers({
    playField,
    partsField,
    selectedShape,
    cursor,
    selectedGroupId,
    player,
    turn,
    alreadyPut,
})
