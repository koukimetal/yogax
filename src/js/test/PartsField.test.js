import {PartsFieldComponent} from '../redux/PartsField';
import * as common from '../common';
import {fromJS} from 'immutable';
import React from "react";
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

let field;
let selectPart;

beforeEach(() => {
    field = fromJS(common.getField(2));
    selectPart = jest.fn();
});


test('test PartsField Component', () => {
    const component = renderer.create(
        <PartsFieldComponent
            player={0}
            field={field}
            selectedGroupId={-2}
            selectPart={selectPart}
        />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('test click', () => {
    const wrapper = shallow(<PartsFieldComponent
        player={0}
        field={field}
        selectedGroupId={-2}
        selectPart={selectPart}
    />);
    expect(wrapper.find('rect').length).toBe(4);
    wrapper.find('rect').first().simulate('click');
    expect(selectPart.mock.calls.length).toBe(1);
});

