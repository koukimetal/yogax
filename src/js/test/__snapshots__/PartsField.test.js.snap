// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PartsField Component 1`] = `
<svg
  height={560}
  width={560}
>
  <rect
    height={20}
    onClick={[Function]}
    style={
      Object {
        "fill": "gray",
        "stroke": "white",
        "strokeDasharray": Array [],
        "strokeWidth": 2,
      }
    }
    width={20}
    x={0}
    y={0}
  />
  <rect
    height={20}
    onClick={[Function]}
    style={
      Object {
        "fill": "gray",
        "stroke": "white",
        "strokeDasharray": Array [],
        "strokeWidth": 2,
      }
    }
    width={20}
    x={20}
    y={0}
  />
  <rect
    height={20}
    onClick={[Function]}
    style={
      Object {
        "fill": "gray",
        "stroke": "white",
        "strokeDasharray": Array [],
        "strokeWidth": 2,
      }
    }
    width={20}
    x={0}
    y={20}
  />
  <rect
    height={20}
    onClick={[Function]}
    style={
      Object {
        "fill": "gray",
        "stroke": "white",
        "strokeDasharray": Array [],
        "strokeWidth": 2,
      }
    }
    width={20}
    x={20}
    y={20}
  />
</svg>
`;
